<a href="https://colab.research.google.com/github/Qubayl90/Module_3_Advanced_Python/blob/main/Advanced_Python_BuiltInFunctions_Map_Filter_Reduce_Recursion_Updated_Lecture_Rangers59_Practice.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Map, Filter, Reduce, Lambda & Recursion

## Tasks Today:

1) <b>Lambda Functions</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Saving to a Variable <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) Multiple Inputs <br>
 &nbsp;&nbsp;&nbsp;&nbsp; d) Passing a Lambda into a Function <br>
 &nbsp;&nbsp;&nbsp;&nbsp; e) Returning a Lambda from a Function <br>
 &nbsp;&nbsp;&nbsp;&nbsp; f) In-Class Exercise #1 <br>
2) <b>Map</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Using Lambda's with Map <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #2 <br>
3) <b>Filter</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Using Lambda's with Filter <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #3 <br>
4) <b>Reduce</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Using Lambda's with Reduce <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #4 <br>
5) <b>Recursion</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Implementing a Base <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Writing a Factorial Function <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #5 <br>
6) <b>Generators & Iterators</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Yield Keyword <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Inifinite Generator <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #6 <br>
7) <b>Exercises</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Exercise #1 - Filtering Empty Strings <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Exercise #2 - Sorting with Last Name <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) Exercise #3 - Conversion to Farhenheit <br>
 &nbsp;&nbsp;&nbsp;&nbsp; d) Exercise #4 - Fibonacci Sequence <br>

## Lambda Functions <br>
<p>Lambda functions... or "Anonymous Functions" are referring to inline functions with no name. The keyword lambda denotes the no name function, and executes within a single line. Without saving it to a variable; however, it is not able to be used, unless passed in either as a paramater or within list comprehension.<br>Written as "(keyword lambda) (one or more inputs) (colon) (function to be executed)"</p>

#### Syntax

In [ ]:
def addTwo(x):
  return x+2

print(addTwo(4))

print((lambda x:x+2)(4))

a_num = 4

print((lambda x:x+2)(a_num))

6
6
6


#### Saving to a Variable

In [ ]:
f_test = lambda x: x+2
f_test(4)

6

#### Multiple Inputs

In [ ]:
print((lambda x,y,z: x * y *z)(3,5,8))

x_test = lambda x,y,z: x*y*z
print(x_test(3,5,8))

120
120


#### Passing a Lambda into a Function

In [ ]:
def multiply(f,num):
  '''
  f expects a lambda function
  num expects a number
  '''
  return f(num)

multiply(lambda x: x*x, 4)

16

#### Returning a Lambda from a Function

In [ ]:
# Regular defined function
def multiply_test(num):
  return num*4
#Function within function
def returnFunc():
  test = 4
  def multiply(num):
    return num * 2
  return multiply

f_return = returnFunc()
print(returnFunc)
print(f_return(4))

#Lambda function returned from a regular function
def returnLam(b,c):
  return lambda x, a: x + a + b + c
r_lamb = returnLam(4,5)   #b and c
print(r_lamb(8,9))        #x and a

<function returnFunc at 0x7f794f73e430>
8
26


#### If Statements within Lambdas

In [ ]:
# Lambda x: True if (condition) else False
f_condition = lambda num: num * 2 if num >20 else num +2

print(f_condition(15))
print(f_condition(20))
print(f_condition(25))

17
22
50


#### In-Class Exercise #1 <br>
<p>Write an anonymous function that cubes the arguments passed in and assign the anonymous function to a variable 'f'.</p>

In [ ]:
f_cube = lambda num: num*num*num
print(f_cube)
print(f_cube(4))
print(f_cube(6))

<function <lambda> at 0x7f794f9055e0>
64
216


## Map <br>
<p>The map function allows you to iterate over an entire list while running a function on each item of the list. This is why the map function works well with lambda's, because it simplifies things and you write less lines of code.<br>The syntax for a map function is "map(function to be used, list to be used)"<br>However, you must be careful, as the map function returns a map object, not a list. To turn it into a list we use the list() type conversion.</p>

#### Syntax

In [ ]:
def func_square(num1, num2):
  if num1<10 and num2 <10:
    return num1**2, num2**2
  else:
    return num1, num2

numbers = [4, 11, 20, 3, 15, 20]
more_numbers = [4,10,3,2,6]

squared_numbers_map = list(map(func_square,numbers,more_numbers))
print(squared_numbers_map)

[(16, 16), (11, 10), (20, 3), (9, 4), (15, 6)]


#### Using Lambda's with Map

In [ ]:
squared_nums_lambda = list(map(lambda x,y: (x**2, y**2) if x<10 and y<10 else (x,y), numbers, more_numbers))
print(squared_nums_lambda)

[(16, 16), (11, 10), (20, 3), (9, 4), (15, 6)]


#### In-Class Exercise #2 <br>
<p>Use the map function to double each number and minus it by one in the list by using a lambda function</p>

In [ ]:
double_nums_minus_one = list(map(lambda x: x*2 - 1, numbers))
print(double_nums_minus_one)

[7, 21, 39, 5, 29, 39]


## Filter() <br>
<p>Filter's are similar to the map function, where you're able to pass a function argument and a list argument and filter out something from the list based on the conditions passed. Similar to the map function, it returns a filter object, so you need to type convert it to a list()</p>

#### Syntax

In [ ]:
names = ['Bob', 'Andy', 'Maxwell', 'Eva', 'Angel']

def ab_names(name):
  if name[0].lower() == 'a':
    return True
  elif name[0].lower() == 'b':
    return True
  else:
    return False

ab_name_list = list(filter(ab_names, names))
print(ab_name_list) 

['Bob', 'Andy', 'Angel']


#### Using Lambda's with Filter()

In [ ]:
a_names_lambda = list(filter(lambda name: True if (name[0].lower() == 'a' or name[0].lower() == 'b')  else False, names))
print(a_names_lambda)

['Bob', 'Andy', 'Angel']


#### In-Class Exercise #3 <br>
<p>Filter out all the numbers that are below the mean of the list.<br><b>Hint: Import the 'statistics' module</b></p>

In [ ]:
from statistics import mean

number_list = [12, 15, 17, 13, 22, 10, 19]
number_list_agv = mean(number_list)

print(number_list_agv)

numbers_below_avg = list(filter(lambda number_below_avg: number_below_avg < mean(number_list), number_list))

print(numbers_below_avg)

15.428571428571429
[12, 15, 13, 10]


## Reduce() <br>
<p>Be very careful when using this function, as of Python 3 it's been moved to the 'functools' library and no longer is a built-in function.<br>The creator of Python himself, says to just use a for loop instead.</p>

#### Syntax

In [ ]:
from functools import reduce

list_1 = [2, 4, 6, 8, 10]

def addNums(num1, num2):
  return num1 + num2

result_add = reduce(addNums, list_1)

print(result_add)

def subtractNums(num1, num2):
  return num1 - num2

result_sub = reduce(subtractNums, list_1)
print(result_sub)

30
-26


#### Using Lambda's with Reduce()

In [ ]:
result_lamb = reduce(lambda x, y: x + y, list_1)
print(result_lamb)

30


#### In-Class Exercise #4 <br>
<p>Use the reduce function to multiply the numbers in the list below together with a lambda function.</p>

In [ ]:
my_list = [1,2,3,4]
result1_lamb = reduce(lambda x, y: x * y, my_list)
print(result1_lamb)

24


## Recursion <br>
<p>Recursion means that a function is calling itself, so it contanstly executes until a base case is reached. It will then push the returning values back up the chain until the function is complete. A prime example of recursion is computing factorials... such that 5! (factorial) is 5*4*3*2*1 which equals 120.</p>

#### Implementing a Base Case

In [ ]:
def addNums(num):
  if num <= 1:
    print("addNums(1) = 1")
    return num
  else:
    print(f'addNums({num}) = {num} + addNums({num-1})')
    return num + addNums(num-1)

addNums(5)

addNums(5) = 5 + addNums(4)
addNums(4) = 4 + addNums(3)
addNums(3) = 3 + addNums(2)
addNums(2) = 2 + addNums(1)
addNums(1) = 1


15

#### Writing a Factorial Function

In [ ]:
def factorial(num):
  if num<=1:
    return 1
  else:
    return num * factorial(num-1)
factorial(5)

120

#### In-Class Exercise #5 <br>
<p>Write a recursive function that subtracts all numbers to the argument given.</p>

In [ ]:
def subNums(num):
  if num <= 1:
    print("subNums(1) = 1")
    return num
  else:
    print(f'subNums({num}) = {num} - subNums({num-1})')
    return num - subNums(num-1)

subNums(10)

subNums(10) = 10 - subNums(9)
subNums(9) = 9 - subNums(8)
subNums(8) = 8 - subNums(7)
subNums(7) = 7 - subNums(6)
subNums(6) = 6 - subNums(5)
subNums(5) = 5 - subNums(4)
subNums(4) = 4 - subNums(3)
subNums(3) = 3 - subNums(2)
subNums(2) = 2 - subNums(1)
subNums(1) = 1


5

## Generators <br>
<p>Generators are a type of iterable, like lists or tuples. They do not allow indexing, but they can still be iterated through with for loops. They are created using functions and the yield statement.</p>

#### Yield Keyword <br>
<p>The yield keyword denotes a generator, it doesn't return so it won't leave the function and reset all variables in the function scope, instead it yields the number back to the caller.</p>

In [1]:
def my_range(stop, start, step = 2):
  while start<stop:
    yield start
    start  = start + step

for i in my_range(20, start = 2):
  my_generator_value = i
  print(my_generator_value)

my_range(20, start=2)

2
4
6
8
10
12
14
16
18


<generator object my_range at 0x7f5092ff2900>

#### Infinite Generator

In [ ]:
# bad, never create infinite loops

#### In-Class Exercise #6 <br>
<p>Create a generator that takes a number argument and yields that number squared, then prints each number squared until zero is reached.</p>

In [3]:
def my_seq_range(stop, start, step = 1):
  while start>stop:
    yield start*start
    start  = start - step

for i in my_seq_range(0, start = 7):
  my_seq_generator_value = i
  print(my_seq_generator_value)

my_seq_range(20, start=2)

49
36
25
16
9
4
1


<generator object my_seq_range at 0x7f5072aa2f90>

# Exercises

### Exercise #1 <br>
<p>Filter out all of the empty strings from the list below</p>

`Output: ['Argentina', 'San Diego', 'Boston', 'New York']`

In [5]:
places = [" ","Argentina", " ", "San Diego","","  ","","Boston","New York"]
not_empty_places = list(filter(lambda non_empty_place: non_empty_place.strip() != "", places))

print(not_empty_places)

['Argentina', 'San Diego', 'Boston', 'New York']


### Exercise #2 <br>
<p>Write an anonymous function that sorts this list by the last name...<br><b>Hint: Use the ".sort()" method and access the key"</b></p>

`Output: ['Victor aNisimov', 'Gary A.J. Bernstein', 'Joel Carter', 'Andrew P. Garfield', 'David hassELHOFF']`

In [4]:
author = ["Joel Carter", "Victor aNisimov", "Andrew P. Garfield","David hassELHOFF","Gary A.J. Bernstein"]
author.sort(key = lambda x: x.split()[-1].upper())
print(author)

['Victor aNisimov', 'Gary A.J. Bernstein', 'Joel Carter', 'Andrew P. Garfield', 'David hassELHOFF']


### Exercise #3 <br>
<p>Convert the list below from Celsius to Farhenheit, using the map function with a lambda...</p>

`Output: [('Nashua', 89.6), ('Boston', 53.6), ('Los Angelos', 111.2), ('Miami', 84.2)]
`

In [3]:
# F = (9/5)*C + 32
places = [('Nashua',32),("Boston",12),("Los Angelos",44),("Miami",29)]

farhenheit_places = list(map(lambda place:(place[0], (place[1]*(9/5)+32)), places))
print(farhenheit_places)

[('Nashua', 89.6), ('Boston', 53.6), ('Los Angelos', 111.2), ('Miami', 84.2)]


### Exercise #4 <br>
<p>Write a recursion function to perform the fibonacci sequence up to the number passed in.</p>

`Output for fib(5) => 
Iteration 0: 1
Iteration 1: 1
Iteration 2: 2
Iteration 3: 3
Iteration 4: 5
Iteration 5: 8`

In [18]:
def fib(n):
  if n == 0 or n==1:
    #print(f'Iteration {n}: {n}')
    return n
  else:
    #fib_value = fib(n-1)+fib(n-2)
    #print(f'Iteration {n}: {fib_value}')
    return (fib(n-1)+fib(n-2))

for i in range(10):
  print(f'Iteration {i}: {fib(i)}')

#print(fib(5))

Iteration 0: 0
Iteration 1: 1
Iteration 2: 1
Iteration 3: 2
Iteration 4: 3
Iteration 5: 5
Iteration 6: 8
Iteration 7: 13
Iteration 8: 21
Iteration 9: 34
